contracts/proposal/dao-proposal-condorcet/src/m.rs

```diff
@@ -0,0 +1,382 @@
+    fn index(&self, (x, y): (u32, u32)) -> u32 {
+        let n = self.n;
+        // the start of the row in `self.cells`.
+        //
+        // the easiest way to conceptualize this is
+        // geometrically. `y*n` is the area of the whole matrix up to
+        // row `y`, and thus the start index of the row if
+        // `self.cells` was not diagonalized [1]. `(y + 1) * y / 2` is the
+        // area of the space that is not in the upper diagonal.
+        //
+        // whole_area - area_of_non_diagonal = area_of_diagonal
+        //
+        // because we're in the land of discrete math and things are
+        // zero-indexed, area_of_diagonal == start_of_row.
+        let row = y * n - (y + 1) * y / 2;
+        // we know that x > y, so to get the index in `self.cells` we
+        // offset x by the distance of x from the line x = y (the
+        // diagonal), as `self.cells`' first index corresponds to the
+        // first item in that row of the upper diagonal.
+        let offset = x - (y + 1);
+        row + offset
+    }
+
+    pub(crate) fn get(&self, (x, y): (u32, u32)) -> Cell {
+        if x < y {
+            self.get((y, x)).invert()
+        } else {
+            let i = self.index((x, y)) as usize;
+            self.cells[i]
+        }
+    }
+
+    pub fn increment(&mut self, (x, y): (u32, u32), amount: Uint128) {
+        debug_assert!(x != y);
+        if x < y {
+            self.decrement((y, x), amount)
+        } else {
+            let i = self.index((x, y)) as usize;
+            self.cells[i] = self.cells[i].increment(amount)
+        }
+    }
+
+    pub fn decrement(&mut self, (x, y): (u32, u32), amount: Uint128) {
+        debug_assert!(x != y);
+        if x < y {
+            self.increment((y, x), amount)
+        } else {
+            let i = self.index((x, y)) as usize;
+            self.cells[i] = self.cells[i].decrement(amount)
+        }
+    }
+
+    /// Computes statistics about M which are used to determine if a
+    /// proposal has passed or may be rejected early.
+    ///
+    /// Code comments refer to this proof of conditions for early
+    /// rejection:
+    ///
+    /// https://github.com/DA0-DA0/dao-contracts/wiki/Proofs-of-early-rejection-cases-for-Condorcet-proposals
+    pub fn stats(&self, power_outstanding: Uint128) -> Stats {
+        let n = self.n;
+        let mut no_winnable_columns = true;
+        for col in 0..n {
+            let mut distance_from_positivity = Uint256::zero();
+            let mut min_margin = Uint128::MAX;
+            let mut max_negative = Uint128::zero();
+            for row in 0..n {
+                if row != col {
+                    match self.get((col, row)) {
+                        Cell::Positive(p) => {
+                            if p < min_margin {
+                                min_margin = p
+                            }
+                        }
+                        Cell::Negative(n) => {
```

Comment:
super nit: want to call this something other than n to avoid confusion
                        // with the above n?

Reply:
lol great point i didn't even realize there were two `n` variables in scope